rethinkdb/datadog_checks/rethinkdb/queries_impl.py

```diff
@@ -0,0 +1,339 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import collections
+import itertools
+from typing import Any, Dict, List, Optional
+
+import rethinkdb
+
+from .utils import to_time_elapsed
+from .version import parse_version
+
+r = rethinkdb.r
+
+# See: https://rethinkdb.com/docs/system-tables/
+SYSTEM = r.db('rethinkdb')
+STATS = SYSTEM.table('stats')
+DB_CONFIG = SYSTEM.table('db_config')
+SERVER_CONFIG = SYSTEM.table('server_config')
+SERVER_STATUS = SYSTEM.table('server_status')
+TABLE_CONFIG = SYSTEM.table('table_config')
+TABLE_STATUS = SYSTEM.table('table_status')
+JOBS = SYSTEM.table('jobs')
+CURRENT_ISSUES = SYSTEM.table('current_issues')
+
+
+def get_cluster_metrics(conn):
+    # type: (rethinkdb.net.Connection) -> List[tuple]
+    """
+    Return cluster-wide metrics.
+    """
+    cluster_counts = r.expr(
+        {
+            'servers': SERVER_CONFIG.count(),
+            'databases': DB_CONFIG.count(),
+        }
+    ).run(conn)
+
+    cluster_stats = STATS.get(['cluster']).run(conn)
+
+    row = (
+        cluster_counts['servers'],
+        cluster_counts['databases'],
+        cluster_stats['query_engine']['queries_per_sec'],
+        cluster_stats['query_engine']['read_docs_per_sec'],
+        cluster_stats['query_engine']['written_docs_per_sec'],
+    )
+
+    return [row]
+
+
+def get_server_metrics(conn):
+    # type: (rethinkdb.net.Connection) -> List[tuple]
+    """
+    Return metrics about each `server`.
+    """
+    # For servers, stats['id'] has the form ['server', '<SERVER_ID>']
+    is_server_stats_row = r.row['id'].nth(0) == 'server'
+    server_id = r.row['id'].nth(1)
+
+    server_statuses = list(SERVER_STATUS.run(conn))
+    joined_server_stats = STATS.filter(is_server_stats_row).eq_join(server_id, SERVER_CONFIG).run(conn)
+
+    rows = []  # type: List[tuple]
+    for document in joined_server_stats:
+        server_stats = document['left']
+        server = document['right']
+        # Join stats and statuses in Python.
+        server_status = next(status for status in server_statuses if status['name'] == server['name'])
+
+        row = (
+            server['name'],
+            # TODO: include `server['tags']` (how?)
```

Comment:
Right now `QueryManager` expects rows to be fixed-size. But in this case we apply a variable amount of tags coming from "server tags" set on RethinkDB startup.

So I think we'll need to add a new feature to `QueryManager`, so it's able to include a list of tags. For example add support for `"type": "tag_list"` columns, which would support reading from a list, or a comma-delimited string.

(Comma delimited string isn't immediately useful for RethinkDB, but could be for future use cases, and just supporting reading from a list would be of very limited use considering `QueryManager` is primarily used with SQL queries.)